driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java

```diff
@@ -387,21 +436,71 @@ private void maybeSpawnNewConnection() {
                 break;
         }
 
-        logger.debug("Creating new connection on busy pool to {}", host);
         manager.blockingExecutor().submit(newConnectionTask);
     }
 
     void replaceDefunctConnection(final PooledConnection connection) {
-        if (connection.markForTrash.compareAndSet(false, true))
+        if (connection.state.compareAndSet(OPEN, GONE))
             open.decrementAndGet();
-        connections.remove(connection);
+        if (connections.remove(connection))
+            manager.blockingExecutor().submit(new Runnable() {
+                @Override
+                public void run() {
+                    addConnectionIfUnderMaximum();
+                }
+            });
         connection.closeAsync();
-        manager.blockingExecutor().submit(new Runnable() {
-            @Override
-            public void run() {
-                addConnectionIfUnderMaximum();
+    }
+
+    void cleanupIdleConnections(long now) {
+        if (isClosed())
+            return;
+
+        shrinkIfBelowCapacity();
+        cleanupTrash(now);
+    }
+
+    /** If we have more active connections than needed, trash some of them */
+    private void shrinkIfBelowCapacity() {
+        int currentLoad = maxTotalInFlight.getAndSet(0);
+
+        int needed = currentLoad / StreamIdGenerator.MAX_STREAM_PER_CONNECTION + 1;
+        if (currentLoad % StreamIdGenerator.MAX_STREAM_PER_CONNECTION > options().getMaxSimultaneousRequestsPerConnectionThreshold(hostDistance))
+            needed += 1;
+        needed = Math.max(needed, options().getCoreConnectionsPerHost(hostDistance));
+        int actual = open.get();
+        int toTrash = Math.max(0, actual - needed);
+
+        logger.trace("Current inFlight = {}, {} connections needed, {} connections available, trashing {}",
+            currentLoad, needed, actual, toTrash);
+
+        if (toTrash <= 0)
+            return;
+
+        for (PooledConnection connection : connections)
+            if (trashConnection(connection)) {
```

Comment:
And another realization:  totalInFlight takes into account inFlight requests on trashed connections a well.  I've seen the following scenario:

* T0:  connections[Connection(0): 128 inFlight,  Connection(1): 1 inFlight], trash[]
* T1:  Connection(1).returnConnection.  connections[Connection(0): 128 inFlight, Connection(1): 0 inFlight], trash[]
* T2:  shrink.  connections[Connection(1); 0 inFlight], trash[Connection(0): 128 inFlight]
* T3: borrowConnection, logic at HostConnectionPool:193 spawns a new connection, takes Connection(0) from Trash, returns Connection(1).  connections[Connection(0): 128 inFlight,  Connection(1): 1 inFlight], trash[]

Because the 128 inFlight requests is taken in count to totalInFlight from Connection(0), Connection(0) gets revived from the trash.  That doesn't seem right does it?    Should we change the logic in HostConnectionPool:193 to only call maybeSpawnNewConnection if leastyBusy active requests == max requests per connection?